modules/network-annotation-processor/src/main/java/org/apache/ignite/internal/network/processor/EnumMethodsGenerator.java

```diff
@@ -0,0 +1,182 @@
+    }
+
+    /**
+     * Collects all {@link Enum}s that are present in network messages without the {@link Transient}.
+     *
+     * @param messageClasses Network message wrappers.
+     */
+    Set<TypeMirror> collectEnums(Collection<MessageClass> messageClasses) {
+        return messageClasses.stream()
+                .flatMap(messageClass -> messageClass.getters().stream())
+                .filter(element -> element.getAnnotation(Transient.class) == null)
+                .map(ExecutableElement::getReturnType)
+                .filter(this::isEnum)
+                .peek(this::checkEnumImplementTransferableEnum)
+                .collect(toSet());
+    }
+
+    /**
+     * Generates a utility class for the {@link Enum} with {@link TransferableEnum}.
+     *
+     * @param enumType {@link Enum} type mirror.
+     * @return Generated class like {@code SomeEnumTransferableUtils}.
+     */
+    TypeSpec generateEnumTransferableUtils(TypeMirror enumType) {
+        assert isEnum(enumType) && isTransferableEnum(enumType) : enumType;
+
+        TypeSpec.Builder typeSpecBuilder = TypeSpec.classBuilder(simpleEnumName(enumType) + "TransferableUtils")
+                .addModifiers(PUBLIC);
+
+        addFormTransferableIdMethod(enumType, typeSpecBuilder);
+
+        return typeSpecBuilder.build();
+    }
+
+    private static void addFormTransferableIdMethod(TypeMirror enumType, TypeSpec.Builder enumTypeSpecBuilder) {
+        ParameterizedTypeName staticFieldType = ParameterizedTypeName.get(
+                ClassName.get(Map.class),
+                TypeName.get(Integer.class),
+                TypeName.get(enumType)
+        );
+
+        FieldSpec staticFieldSpec = FieldSpec.builder(staticFieldType, VALUE_BY_TRANSFERABLE_ID_FIELD_NAME)
+                .addModifiers(PRIVATE, STATIC, FINAL)
+                .build();
+
+        CodeBlock staticBlock = CodeBlock.builder()
+                .beginControlFlow("try")
+                .addStatement("$T[] values = $T.values()", enumType, enumType)
+                .addStatement("var tmp = new $T<$T, $T>()", HashMap.class, Integer.class, enumType)
+                .add("\n")
+                .beginControlFlow("for ($T value : values)", enumType)
+                .addStatement("int $L = value.transferableId()", TRANSFERABLE_ID_FIELD_NAME)
+                .addStatement("assert $L >= 0 : \"Must not be negative: \" + value", TRANSFERABLE_ID_FIELD_NAME)
+                .add("\n")
+                .addStatement("$T prev = tmp.put($L, value)", enumType, TRANSFERABLE_ID_FIELD_NAME)
+                .addStatement("assert prev == null : \"Must be unique: \" + value")
+                .endControlFlow()
+                .add("\n")
+                .addStatement("$L = Map.copyOf(tmp)", VALUE_BY_TRANSFERABLE_ID_FIELD_NAME)
+                .nextControlFlow("catch($T t)", Throwable.class)
+                .addStatement("throw new $T(t)", ExceptionInInitializerError.class)
```

Comment:
Why do we need to wrap everything in `ExceptionInInitializerError`? Would it be enough if the original exception flies away?

Reply:
apply it

Reply:
I use `java.lang.ExceptionInInitializerError` according to the documentation, that an unexpected exception occurred in a static block, I think this is correct, plus according to the project we already use such an exception.